File: src/test/java/tech/sirwellington/alchemy/arguments/assertions/CollectionAssertionsTest.java
package tech.sirwellington.alchemy.arguments.assertions;

import java.util.*;

import org.junit.jupiter.api.RepeatedTest;
import tech.sirwellington.alchemy.arguments.AlchemyAssertion;
import tech.sirwellington.alchemy.arguments.Arguments;
import tech.sirwellington.alchemy.test.AlchemyTest;
import tech.sirwellington.alchemy.test.generation.GenerateList;
import tech.sirwellington.alchemy.test.generation.GenerateMap;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.notNullValue;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static tech.sirwellington.alchemy.arguments.TestHelpers.assertThrowsFailedAssertion;
import static tech.sirwellington.alchemy.arguments.TestHelpers.randomElementFrom;
import static tech.sirwellington.alchemy.arguments.assertions.CollectionAssertions.*;
import static tech.sirwellington.alchemy.test.ThrowableAssertion.assertThrows;

@AlchemyTest
class CollectionAssertionsTest {
    private static final int TEST_ITERATIONS = 50;
    
    private Set<String> emptySet = Collections.emptySet();
    private List<String> emptyList = Collections.emptyList();

    @GenerateList(String.class)
    private List<String> strings;

    @GenerateMap(keyType = String.class, valueType = String.class)
    private Map<String, String> map;

    @RepeatedTest(TEST_ITERATIONS)
    public void testNonEmptyCollection() {
        var instance = CollectionAssertions.<String>nonEmptyCollection();
        assertThat(instance, notNullValue());

        instance.check(strings);
        assertThrowsFailedAssertion(() -> instance.check(null));
        assertThrowsFailedAssertion(() -> instance.check(Collections.emptySet()));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testNonEmptyList() {
        AlchemyAssertion<List<String>> instance = nonEmptyList();
        assertThat(instance, notNullValue());

        instance.check(strings);
        assertThrowsFailedAssertion(() -> instance.check(null));
        assertThrowsFailedAssertion(() -> instance.check(emptyList));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testNonEmptySet() {
        AlchemyAssertion<Set<String>> instance = nonEmptySet();
        assertThat(instance, notNullValue());

        var setOfStrings = Set.copyOf(strings);
        instance.check(setOfStrings);

        assertThrowsFailedAssertion(() -> instance.check(new HashSet<>()));
        assertThrowsFailedAssertion(() -> instance.check(null));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testNonEmptyMap() {
        AlchemyAssertion<Map<String,Integer>> instance = nonEmptyMap();
        assertThat(instance, notNullValue());

        var map = new HashMap<String, Integer>();
        for (int i = 0; i < 5; i++) {
            map.put("key" + i, i);
        }
        instance.check(map);

        assertThrowsFailedAssertion(() -> instance.check(Collections.emptyMap()));
        assertThrowsFailedAssertion(() -> instance.check(null));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testNonEmptyArray() {
        AlchemyAssertion<String[]> instance = nonEmptyArray();
        assertThat(instance, notNullValue());

        var array = strings.toArray(new String[0]);
        instance.check(array);

        assertThrowsFailedAssertion(() -> instance.check(null));
        assertThrowsFailedAssertion(() -> instance.check(new String[0]));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testListContaining() {
        // Given
        var element = randomElementFrom(strings);
        var instance = listContaining(element);
        // Then
        Tests.checkForNullCase(instance);

        // Then
        instance.check(strings);
        // Then
        assertThrowsFailedAssertion(
            () -> instance.check(Collections.singletonList("xyz"))
        );
        assertThrowsFailedAssertion(
            () -> instance.check(Collections.emptyList())
        );
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testListContainingWithBadArgs() {
        assertThrows(() -> listContaining(null));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContaining() {
        var element = randomElementFrom(strings);
        var instance = collectionContaining(element);
        assertThat(instance, notNullValue());

        instance.check(strings);
        assertThrowsFailedAssertion(() -> instance.check(Collections.singletonList("xyz")));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContainingWithBadArgs() {
        assertThrows(() -> collectionContaining(null)).isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContainingAll() {
        var args = strings.subList(0, 2);
        var collection = new ArrayList<>(strings);
        collection.addAll(args);

        var first = args.getFirst();
        var others = args.subList(1, args.size()).toArray(new String[0]);

        var instance = collectionContainingAll(first, others);
        assertThat(instance, notNullValue());

        instance.check(collection);
        assertThrowsFailedAssertion(() -> instance.check(Collections.singletonList("xyz")));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContainingAllWithBadArgs() {
        assertThrows(() -> collectionContainingAll((String[]) null))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContainingAtLeastOneOf() {
        var args = strings.subList(0, 2);
        var collection = new ArrayList<>(strings);
        collection.addAll(args);

        var first = args.getFirst();
        var others = args.subList(1, args.size()).toArray(new String[0]);

        var instance = collectionContainingAtLeastOneOf(first, others);
        assertThat(instance, notNullValue());

        instance.check(collection);
        assertThrowsFailedAssertion(() -> instance.check(Collections.singletonList("xyz")));

        // Re-add first to make it pass
        collection.add(first);
        instance.check(collection);
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionContainingAtLeastOneOfWithBadArgs() {
        assertThrows(() -> collectionContainingAtLeastOneOf((String[]) null))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testMapWithKey() {
        // Given
        var map = new HashMap<String, String>();
        for (int i = 0; i < 5; i++) {
            map.put("key" + i, "val" + i);
        }
        var key = map.keySet().iterator().next();

        var instance = CollectionAssertions.<String, String>mapWithKey(key);
        assertThat(instance, notNullValue());

        instance.check(map);
        assertThrowsFailedAssertion(() -> instance.check(Collections.singletonMap("bad", "val")));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testMapWithKeyValue() {
        var map = new HashMap<String, Integer>();
        
        for (int i = 0; i < 5; i++) {
            map.put(String.valueOf(i), i);
        }

        Map.Entry<String, Integer> entry = map.entrySet().iterator().next();
        var instance = mapWithKeyValue(entry.getKey(), entry.getValue());
        assertThat(instance, notNullValue());

        instance.check(map);
        assertThrowsFailedAssertion(() -> instance.check(Collections.singletonMap("bad", -1)));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testKeyInMap() {
        var map = new HashMap<String, String>();
        for (int i = 0; i < 5; i++) {
            map.put("key" + i, "val" + i);
        }
        var assertion = keyInMap(map);

        var anyKey = map.keySet().iterator().next();
        assertion.check(anyKey);

        assertThrowsFailedAssertion(() -> assertion.check("nonexistent"));
        assertThrowsFailedAssertion(() -> assertion.check(null));

        assertThrows(() -> keyInMap((Map<?, ?>) null))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testKeyInMapWithEmptyMap() {
        var assertion = keyInMap(Collections.emptyMap());

        for (String s : strings) {
            assertThrowsFailedAssertion(() -> assertion.check(s));
        }
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testValueInMap() {
        Map<String, String> map = new HashMap<>();
        for (int i = 0; i < 5; i++) {
            map.put("key" + i, "val" + i);
        }
        var assertion = valueInMap(map);

        String anyValue = map.values().iterator().next();
        assertion.check(anyValue);

        assertThrowsFailedAssertion(() -> assertion.check("nonexistent"));
        assertThrowsFailedAssertion(() -> assertion.check(null));

        assertThrows(() -> valueInMap((Map<?, ?>) null))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testValueInMapWithEmptyMap() {
        var assertion = valueInMap(Collections.emptyMap());

        for (String s : strings) {
            assertThrowsFailedAssertion(() -> assertion.check(s));
        }
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testElementInCollection() {
        var assertion = elementInCollection(strings);

        var anyValue = strings.get(Math.abs(new Random().nextInt()) % strings.size());
        assertion.check(anyValue);

        assertThrowsFailedAssertion(() -> assertion.check("nonexistent"));
        assertThrowsFailedAssertion(() -> assertion.check(null));

        assertThrows(() -> elementInCollection((List<?>) null))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionOfSize() {
        // Given
        int size = strings.size();
        var instance = CollectionAssertions.<String, Collection<String>>collectionOfSize(size);

        // Then
        assertDoesNotThrow(() -> instance.check(strings));

        // When
        strings.add("extra");

        // Then
        assertThrowsFailedAssertion(() -> instance.check(strings));

        // Then
        var newSize = strings.size();
        assertDoesNotThrow(
            () -> Arguments.checkThat(strings).isA(collectionOfSize(newSize))
        );
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testCollectionOfSizeWithBadArgs() {
        int badSize = -1;
        assertThrows(() -> collectionOfSize(badSize))
            .isIllegalArgumentException();
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testEmptyCollection() {
        // Given
        var instance = CollectionAssertions.<String,Collection<String>>emptyCollection();
        // Then
        assertThat(instance, notNullValue());

        // Then
        assertDoesNotThrow(
            () -> instance.check(Collections.emptySet())
        );
        assertThrowsFailedAssertion(() -> instance.check(strings));
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testEmptyList() {
        // When
        var instance = emptyList();
        // Then
        assertThat(instance, notNullValue());

        // Then
        assertDoesNotThrow(
            () -> instance.check(Collections.emptyList())
        );
        assertThrowsFailedAssertion(
            () -> Arguments.checkThat(strings).is(emptyList())
        );
    }

    @RepeatedTest(TEST_ITERATIONS)
    public void testEmptySet() {
        // When
        var instance = emptySet();
        // Then
        assertThat(instance, notNullValue());

        assertDoesNotThrow(
            () -> Arguments.checkThat(emptySet).isA(emptySet())
        );
        var nonEmpty = Set.copyOf(strings);
        assertThrowsFailedAssertion(
            () -> Arguments.checkThat(nonEmpty).isA(emptySet())
        );
    }
}

File: src/test/java/tech/sirwellington/alchemy/arguments/assertions/CollectionAssertionsTest.kt
/*
 * Copyright 2015 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.arguments.assertions

import org.hamcrest.Matchers.notNullValue
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import tech.sirwellington.alchemy.arguments.AlchemyAssertion
import tech.sirwellington.alchemy.arguments.Arguments.checkThat
import tech.sirwellington.alchemy.arguments.FailedAssertionException
import tech.sirwellington.alchemy.test.junit.ThrowableAssertion.assertThrows
import tech.sirwellington.alchemy.test.junit.runners.AlchemyTestRunner
import tech.sirwellington.alchemy.test.junit.runners.DontRepeat
import tech.sirwellington.alchemy.test.junit.runners.GenerateList
import tech.sirwellington.alchemy.test.junit.runners.Repeat
import java.util.*

/**

 * @author SirWellington
 */
@RunWith(AlchemyTestRunner::class)
@Repeat(50)
class CollectionAssertionsTest
{

    @GenerateList(String::class)
    private val strings: MutableList<String>? = null

    @Before
    fun setUp()
    {
    }

    @DontRepeat
    @Test
    fun testCannotInstantiateClass()
    {
        assertThrows { CollectionAssertions::class.java.newInstance() }

        assertThrows { CollectionAssertions() }
                .isInstanceOf(IllegalAccessException::class.java)
    }

    @Test
    fun testNonEmptyCollection()
    {

        val instance = CollectionAssertions.nonEmptyCollection<String>()
        assertThat(instance, notNullValue())

        instance.check(strings)

        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(emptySet<Any>()) }
                .isInstanceOf(FailedAssertionException::class.java)

    }

    @Test
    fun testNonEmptyList()
    {
        val instance = CollectionAssertions.nonEmptyList<String>()
        assertThat(instance, notNullValue())

        instance.check(strings)

        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(emptyList<Any>()) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testNonEmptySet()
    {
        val instance = CollectionAssertions.nonEmptySet<String>()
        assertThat(instance, notNullValue())

        val setOfStrings = HashSet(strings!!)
        instance.check(setOfStrings)

        checkThat<Set<String>>(setOfStrings).`is`(Companion.nonEmptySet())

        val emptySet = HashSet<String>()
        assertThrows { instance.check(emptySet) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testNonEmptyMap()
    {
        val instance = CollectionAssertions.nonEmptyMap<String, Int>()

        val map = mapOf<String, Int>(alphabeticString(),
                                     positiveIntegers(),
                                     40)

        instance.check(map)

        assertThrows { instance.check(emptyMap<Any, Any>()) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testNonEmptyArray()
    {
        val instance = CollectionAssertions.nonEmptyArray<String>()
        assertThat(instance, notNullValue())

        assertThrows { instance.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        val stringArray = strings!!.toTypedArray()

        instance.check(stringArray)

        val emptyStringArray = arrayOf<String>()

        assertThrows { instance.check(emptyStringArray) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testListContaining()
    {
        val string = strings!!.stream().findAny().orElseGet(alphabeticString())

        val instance = CollectionAssertions.listContaining(string)
        assertThat(instance, notNullValue())
        instance.check(strings)

        val hex = listOf<String>(hexadecimalString(100))

        assertThrows { instance.check(hex) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun testListContainingWithBadArgs()
    {
        assertThrows { CollectionAssertions.listContaining<Any>(null) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testCollectionContaining()
    {
        val string = strings!!.stream().findAny().orElseGet(alphabeticString())

        val instance = CollectionAssertions.collectionContaining<Any, Collection<String>>(string)
        assertThat(instance, notNullValue())
        instance.check(strings)

        val hex = listOf<String>(hexadecimalString(100))

        assertThrows { instance.check(hex) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testCollectionContainingWithBadArgs()
    {
        assertThrows { CollectionAssertions.collectionContaining<Any, Collection<*>>(null) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun testCollectionContainingAll()
    {
        val args = listOf<String>(alphabeticString())
        val collection = listOf<String>(alphabeticString())
        collection.addAll(args)

        val first = args[0]
        val others = args.subList(1, args.size).toTypedArray()

        val instance = CollectionAssertions.collectionContainingAll<Any, Collection<String>>(first, *others)
        assertThat(instance, notNullValue())
        instance.check(collection)

        val otherCollection = listOf<String>(alphabeticString())
        assertThrows { instance.check(otherCollection) }.isInstanceOf(FailedAssertionException::class.java)
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testCollectionContainingAllWithBadArgs()
    {
        assertThrows { CollectionAssertions.collectionContainingAll<Any, Collection<*>>(null) }.isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun testCollectionContainingAtLeastOnceOf()
    {
        val args = listOf<String>(alphanumericString())
        val collection = listOf<String>(alphanumericString())
        collection.addAll(args)

        val first = collection[0]
        val others = args.subList(1, args.size).toTypedArray()

        val instance = CollectionAssertions.collectionContainingAtLeastOnceOf<Any, Collection<String>>(first, *others)
        instance.check(collection)

        val otherCollection = listOf<String>(alphabeticString())
        assertThrows { instance.check(otherCollection) }

        //With at least one, it should pass.
        otherCollection.add(first)
        instance.check(otherCollection)
    }

    @DontRepeat
    @Test
    @Throws(Exception::class)
    fun testCollectionContainingAtLeastOnceOfWithBadArgs()
    {
        assertThrows { CollectionAssertions.collectionContainingAtLeastOnceOf<Any, Collection<*>>(null) }.isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testMapWithKey()
    {
        val map = mapOf<Int, String>(positiveIntegers(), hexadecimalString(100), 100)

        val key = map.keys.stream().findAny().get()

        val instance = CollectionAssertions.mapWithKey<Int, String>(key)
        assertThat(instance, notNullValue())

        instance.check(map)

        val badMap = mapOf<Int, String>(negativeIntegers(), hexadecimalString(100), 100)
        assertThrows { instance.check(badMap) }
    }

    @Test
    fun testMapWithKeyValue()
    {
        val map = mapOf<Int, String>(positiveIntegers(), alphabeticString(), 100)

        val anyEntry = map.entries.stream().findAny().get()

        val instance: AlchemyAssertion<Map<Int, String>>
        instance = CollectionAssertions.mapWithKeyValue(anyEntry.key, anyEntry.value)
        assertThat(instance, notNullValue())

        //Should pass OK
        instance.check(map)

        val badMap = mapOf<Int, String>(negativeIntegers(), hexadecimalString(100), 100)
        assertThrows { instance.check(badMap) }
    }

    @Test
    fun testKeyInMap()
    {
        val map = mapOf<String, String>(alphabeticString(), alphanumericString(), 25)

        val assertion = CollectionAssertions.keyInMap(map)
        assertThat(assertion, notNullValue())

        val anyKey = map.keys.stream().findAny().get()
        assertion.check(anyKey)

        val randomKey = one(hexadecimalString(42))
        assertThrows { assertion.check(randomKey) }
                .isInstanceOf(FailedAssertionException::class.java)

        //Edge cases
        assertThrows { assertion.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { CollectionAssertions.keyInMap<Any, Any>(null!!) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testKeyInMapWithEmptyMap()
    {
        val assertion = CollectionAssertions.keyInMap(emptyMap<Any, Any>())

        for (string in strings!!)
        {
            assertThrows { assertion.check(string) }
                    .isInstanceOf(FailedAssertionException::class.java)
        }
    }

    @Test
    fun testValueInMap()
    {

        val map = mapOf<String, String>(alphanumericString(), alphabeticString(), 24)

        val assertion = CollectionAssertions.valueInMap(map)
        assertThat(assertion, notNullValue())

        val anyValue = map.values.stream().findAny().get()
        assertion.check(anyValue)

        val randomValue = one(hexadecimalString(10))
        assertThrows { assertion.check(randomValue) }
                .isInstanceOf(FailedAssertionException::class.java)

        //Edge cases
        assertThrows { assertion.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { CollectionAssertions.valueInMap<Any, Any>(null!!) }
                .isInstanceOf(IllegalArgumentException::class.java)

        //Empty map should be ok
        CollectionAssertions.valueInMap(emptyMap<Any, Any>())
    }

    @Test
    fun testValueInMapWithEmptyMap()
    {
        val assertion = CollectionAssertions.valueInMap(emptyMap<Any, Any>())
        assertThat(assertion, notNullValue())

        for (string in strings!!)
        {
            assertThrows { assertion.check(string) }
                    .isInstanceOf(FailedAssertionException::class.java)
        }
    }

    @Test
    fun testElementInCollection()
    {
        val assertion = CollectionAssertions.elementInCollection(strings!!)
        assertThat(assertion, notNullValue())

        val anyValue = strings.stream().findAny().get()
        assertion.check(anyValue)

        val randomValue = one(hexadecimalString(20))

        assertThrows { assertion.check(randomValue) }
                .isInstanceOf(FailedAssertionException::class.java)

        //Edge cases
        assertThrows { assertion.check(null) }
                .isInstanceOf(FailedAssertionException::class.java)

        assertThrows { CollectionAssertions.elementInCollection<Any>(null!!) }
                .isInstanceOf(IllegalArgumentException::class.java)

        //Empty Collections should be ok
        CollectionAssertions.elementInCollection(emptyList<Any>())

    }

    @Test
    fun testCollectionOfSize()
    {
        val size = strings!!.size
        val instance = CollectionAssertions.collectionOfSize<in Collection<String>>(size)
        instance.check(strings)

        strings.add(one(alphabeticString()))

        assertThrows { instance.check(strings) }
                .isInstanceOf(FailedAssertionException::class.java)

        checkThat<List<String>>(strings)
                .`is`(Companion.collectionOfSize(strings.size))

    }

    @DontRepeat
    @Test
    fun testCollectionOfSizeWithBadArgs()
    {
        val badSize = one(negativeIntegers())

        assertThrows { CollectionAssertions.collectionOfSize<Collection<*>>(badSize) }
                .isInstanceOf(IllegalArgumentException::class.java)
    }

    @Test
    fun testEmptyCollection()
    {
        val instance = CollectionAssertions.emptyCollection<String>()

        val emptyCollection = ArrayList<String>()

        instance.check(emptyCollection)

        assertThrows { instance.check(strings) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testEmptyList()
    {
        val instance = CollectionAssertions.emptyList<String>()

        val emptyList = LinkedList<String>()
        instance.check(emptyList)

        assertThrows { instance.check(strings) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

    @Test
    fun testEmptySet()
    {
        val instance = CollectionAssertions.emptySet<String>()

        val emptySet = HashSet<String>()
        instance.check(emptySet)

        val nonEmptySet = HashSet(strings!!)
        assertThrows { instance.check(nonEmptySet) }
                .isInstanceOf(FailedAssertionException::class.java)
    }

}
